File: archive/nodejs/muse/algorithms/number_theory/Primality.js
'use strict'

exports.isPrime = (n) => {
  if (n < 2) {
    return false
  }
  if ((n & 1) === 0 && n !== 2) {
    return false
  }

  for (let i = 3, bound = Math.floor(Math.sqrt(n)) + 1; i < bound; i += 2) {
    if (n % i === 0) {
      return false
    }
  }

  return true
}

exports.isComposite = (n) => {
  return n > 1 && !this.isPrime(n)
}

File: archive/ruby/muse/algorithms/number_theory/Primality.rb
module Primality
  def self.isPrime(n)
    return false if n < 2
    return false if (n & 1).zero? && n != 2

    3.step(Integer.sqrt(n), 2) do |i|
      return false if (n % i).zero?
    end

    true
  end

  def self.isComposite(n)
    n > 1 && !isPrime(n)
  end
end
